### PR TITLE
Zebra does not properly track which route-maps are changed (#2493)

### PR DESCRIPTION
**Pull request by kssoman**

### CODE REVIEW

**Comment by kssoman** [comment]
Test logs
[newlogs.txt](https://github.com/kssoman/frr/files/2267376/newlogs.txt)

File: zebra/redistribute.c [hunk 1]
```diff
@@ -715,7 +715,7 @@ int zebra_import_table_config(struct vty *vty)
 	return write;
 }
 
-void zebra_import_table_rm_update()
+void zebra_import_table_rm_update(char *rmap)
```
Comment (srimohans) [inline zebra/redistribute.c:718]: Starting Review

File: zebra/zebra_rib.c [hunk 2]
```diff
@@ -2707,12 +2706,18 @@ void rib_update(vrf_id_t vrf_id, rib_update_event_t event)
 
 	/* Process routes of interested address-families. */
 	table = zebra_vrf_table(AFI_IP, SAFI_UNICAST, vrf_id);
-	if (table)
+	if (table) {
```
Comment (srimohans) [inline zebra/zebra_rib.c:2723]: There seems to be an issue with tab/spaces on your computer. It is changing the line alignment and this issue will come back in external review. Could you please check this ?

File: zebra/zebra_routemap.c [hunk 3]
```diff
@@ -1289,6 +1322,9 @@ static void zebra_route_map_process_update_cb(char *rmap_name)
 	if (IS_ZEBRA_DEBUG_EVENT)
 		zlog_debug("Event handler for route-map: %s",
 			   rmap_name);
+	zebra_import_table_rm_update(rmap_name);
```
Comment (srimohans) [inline zebra/zebra_routemap.c]: 1) I do not understand on why these 4 lines of code was moved from update_timer function to here. 
2) Don't think this is a right function, as function name and the new code will not be related. Lets discuss

File: zebra/zebra_routemap.c [hunk 4]
```diff
@@ -1066,6 +1065,39 @@ static struct route_map_rule_cmd route_match_ip_address_prefix_list_cmd = {
 	route_match_ip_address_prefix_list_free};
 
 
+static route_map_result_t
```
Comment (srimohans) [inline zebra/zebra_routemap.c]: This code changes seems to big, Can you explain in an email about these changes ? Also, can you add one local reviewer from India team ?

**Review by srimohans — CHANGES_REQUESTED** [review]

**Comment by kssoman** [comment]
Description of changes 
--------------------------  

There are 3 types of routemaps in zebra :  

RIB routemap (for filtering routes in RIB)  :  ip/ipv6 protocol <name> route-map <name>

NHT routemap (for filtering routes in nexthop table)  : ip/ipv6 nht <protocol>  route-map <name>

Import table route map  :  ip import-table (1-252) [distance (1-255)] [route-map WORD]

The routemap code in zebra does not check which routemap changed and calls functions for all when any routemap changes. The fix is to find the changed routemap in the appropriate table and call only relevant functions like : 
rib_update()  :  RIB routemap change
zebra_evaluate_rnh() :  NHT routemap change
zebra_import_table_rm_update()  :  Import routemap change

Change in routemap involves adding new clause ,  change to prefix list, access list etc.

The routemap mappings are stored in the following :   ( File : zebra_routemap.c)
This is set by configuration commands mentioned above

char *proto_rm[AFI_MAX][ZEBRA_ROUTE_MAX + 1]; /* "any" == ZEBRA_ROUTE_MAX */
/* NH Tracking route map */
char *nht_rm[AFI_MAX][ZEBRA_ROUTE_MAX + 1]; /* "any" == ZEBRA_ROUTE_MAX */
char *zebra_import_table_routemap[AFI_MAX][ZEBRA_KERNEL_TABLE_MAX];

When any routemap changes,  zebra_route_map_update_timer() is started :  zebra_route_map_mark_update() -> zebra_route_map_update_timer()

This walks through the routemap list and processes the changed routemaps : route_map_walk_update_list( zebra_route_map_process_update_cb)

In the function zebra_route_map_update_timer(), currently all functions for all types are routemaps are called without checking which routemap changed : 
The function zebra_evaluate_rnh() need not be called for RIB routemap change
The function rib_update() need not be called for NHT routemap change

Therefore the functions are moved to zebra_route_map_process_update_cb(char *rmap_name)
This will call : 
zebra_import_table_rm_update(rmap_name)
zebra_rib_table_rm_update(rmap_name)
zebra_nht_rm_update(rmap_name)

These function will verify if the changed routemap (specified by the parameter)  is related for that table by matching with the configured table mappings as mentioned above. If the match is found
the appropriate functions are called.  This prevents calling functions not related to a particular routemap

Also added code for "match ipv6 address prefix list <name>"  which is currently not supported.

**Pull request merged**